ctrd/container.go

```diff
@@ -500,3 +500,27 @@ func (c *Client) ResizeContainer(ctx context.Context, id string, opts types.Resi
 
 	return pack.task.Resize(ctx, uint32(opts.Height), uint32(opts.Width))
 }
+
+// WaitContainer waits until container's status is stopped.
+func (c *Client) WaitContainer(ctx context.Context, id string) (int64, error) {
+	wrapperCli, err := c.Get(ctx)
+	if err != nil {
+		return -1, fmt.Errorf("failed to get a containerd grpc client: %v", err)
+	}
+
+	ctx = leases.WithLease(ctx, wrapperCli.lease.ID())
+
+	waitExit := func() *Message {
+		return c.ProbeContainer(ctx, id, -1*time.Second)
```

Comment:
I still have concerns about this function `ProbeContainer`, because it consumes the channel receiver's data. If you consumes the data here, some function will loss the data.

Reply:
@fuweid You're right. But this channel `ch` seems only to be consumed in function `ProbeContainer`. And this function is called in function `DestroyContainer` in order to wait container exits normally. And if you don't mind, I can add additional channel `waitMeg        chan *Message` in struct `containerPack` for `WaitContainer` function to consume specially. I don't have a more elegant way to solve it out...

Reply:
Hi @xiechengsheng , please go ahead and show your code. We can find the way out. 😄

Reply:
@xiechengsheng @fuweid ProbeContainer consumes the channel data, and it will push the data back again. So we will not loss the data

Reply:
Solved